File: src/subforge/models/project.py
"""Canonical project data model — the single source of truth (ARCH §16, §17)."""

from enum import Enum

from pydantic import BaseModel, Field


class StageState(str, Enum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    SKIPPED = "SKIPPED"


class Segment(BaseModel):
    """Canonical caption unit. Timestamps are seconds (float)."""

    id: int
    start: float
    end: float
    source: str
    speaker: str | None = None
    translations: dict[str, str] = Field(default_factory=dict)


class TranscriptSegment(BaseModel):
    id: int
    start: float
    end: float
    text: str
    speaker: str | None = None


class Transcript(BaseModel):
    """Normalized ASR output — identical regardless of provider (ARCH §6)."""

    language: str | None = None
    segments: list[TranscriptSegment]


class ProjectMeta(BaseModel):
    name: str
    source_language: str
    target_languages: list[str] = Field(default_factory=list)
    speaker_map: dict[str, str] = Field(default_factory=dict)


class Project(BaseModel):
    project: ProjectMeta
    segments: list[Segment]
    stages: dict[str, StageState] = Field(default_factory=dict)

    def get_stage(self, name: str) -> StageState:
        return self.stages.get(name, StageState.PENDING)

    def set_stage(self, name: str, state: StageState) -> None:
        self.stages[name] = state

File: internal/domain/project.go
package domain

import "time"

type StageStatus string

const (
	StatusPending   StageStatus = "pending"
	StatusRunning   StageStatus = "running"
	StatusCompleted StageStatus = "completed"
	StatusFailed    StageStatus = "failed"
	StatusSkipped   StageStatus = "skipped"
)

type Segment struct {
	ID      int     `json:"id"`
	Start   float64 `json:"start"`
	End     float64 `json:"end"`
	Source  string  `json:"source"`
	Speaker string  `json:"speaker,omitempty"`
}

type Project struct {
	Name          string                 `json:"name"`
	AudioPath     string                 `json:"audio_path"`
	AudioDuration float64                `json:"audio_duration,omitempty"`
	Language      string                 `json:"language"`
	Model         string                 `json:"model"`
	Stages        map[string]StageStatus `json:"stages"`
	Error         string                 `json:"error,omitempty"`
	Segments      []Segment              `json:"segments"`
	CreatedAt     time.Time              `json:"created_at"`
	UpdatedAt     time.Time              `json:"updated_at"`
}

func NewProject(name, audioPath, model, language string) *Project {
	now := time.Now().UTC()
	if language == "" {
		language = "auto"
	}
	if model == "" {
		model = "small"
	}
	return &Project{
		Name:      name,
		AudioPath: audioPath,
		Language:  language,
		Model:     model,
		Stages: map[string]StageStatus{
			"transcribe": StatusPending,
			"export":     StatusPending,
		},
		Segments:  make([]Segment, 0),
		CreatedAt: now,
		UpdatedAt: now,
	}
}
